### DISCUSSION TITLE
Transportation - connectors_ids vs segment's geometry

### DISCUSSION THREAD

**Discussion by Azbesciak** [OvertureMaps/data#284] [Discussion]
Hi, is it said that every connector needs to have a reference in a segment geometry - 1:1, I mean the same coordinate is mentioned in the geometry? Or I need to place them somehow in the geometry? If so - how to manage index of that connector in the geometry?

Is there any order/direction at least assumed?

**Comment by Azbesciak** [comment]
> Note that in the above example, an official Overture Maps data release would insert coordinates in the middle of the north-south segments, if they did not already exist, because Overture Maps data releases will always ensure that **every segment's geometry includes all of its connectors**. From a computer's perspective, this is a very minor alteration of the segment's shape.

https://docs.overturemaps.org/schema/attributes/transportation/shape-connectivity/#interior-connectors

**Comment by brad-richardson** [comment]
Looks like you found part of your answer re: coordinates in segment geometries.

Connector ids are assumed to be ordered by relative location from start to end of the segment. We're currently evaluating including a linear reference value (0-1) to indicate understood position for each segment to avoid some ambiguity, let me know if that would be helpful.

**Reply by Azbesciak** [reply]
For the data perspective it would be helpful for sure, but from the other side maybe just enough is to add that info in the schema, in the field description. 
Or maybe not a linear reference value, but the geometry index - that would also make it less complex for the processing. Of course, somehow we have a double source of truth (connectors' geometry and index there, not to mention it may somehow be different across multiple segments, but still - read model vs write model, that would be optimized for reading) but also comparing geometries may cause questions like "is that geometry there" or "should I make some area based equality". Not to mention question "is that coordinate in my segment's geometry". I know the official release promises that, but would be good to have it standardized.

**Comment by Azbesciak** [comment]
Hi, I see you have added `at` property to the segment. As I understand adding `index` was rejected and not planned?

BTW what with the altitude on geometry? I wonder how that `at` is calculated.

**Reply by Azbesciak** [reply]
I am just testing a small part of the data, earlier I needed to calculate distances between geometry points on segments, now I wanted to utilize that to extract indices of these connectors, on geometry.

And I have a problem, you know.

Not to mention, that `at` is with 2 decimal points precision, which may be ok (or not - I saw a couple of geometries which have points nearly overlapping each other), then I really do not know what calculation method was used.

I used haversine (big circle) distance. And for example, I got the following.
connectorId=08f1f1824197146a0477dec08a4781c3,
connectorAt=0.39,
segmentId=0861f18247ffffff04739ed1f25f7f5e,
accumulated offset from segment start, for each geometry point=[0.0, 0.17210651560269796, **0.35235343733315067, 0.43113432389208406**, 0.7232745665830695, 0.8855132339358178]

Regarding the precision, take a look at https://explore.overturemaps.org/#20.53/53.3900054/14.6445999 segment 0871f0e79bffffff04762679d43bfad7
![image](https://github.com/user-attachments/assets/fa3f3b42-b919-43f8-bff7-f1b510b550ea)
0 = {ConnectorElement@33584} "ConnectorElement(at=0.0, connectorID=08f1f0e79b876c68047427d08e9b18a5)"
1 = {ConnectorElement@33585} "ConnectorElement(at=0.4, connectorID=08f1f0e79b84e85a047ea768e34fc13c)"
2 = {ConnectorElement@33586} "ConnectorElement(at=0.4, connectorID=08f1f0e79b84eb9a047cff6afb73af0f)"
3 = {ConnectorElement@33587} "ConnectorElement(at=1.0, connectorID=08f1f0e79bb898990475ed07761a5fb9)"

You know, practically everyone needs to calculate distances anyway. It would be better to just have indices, instead of offset from the geometry start.

**Reply by brad-richardson** [reply]
I think you might be missing spherical projection during calculation?

Here's what I get using distances calculated with projected geometries:
`[{idx=1, at=0.0}, {idx=2, at=0.16}, {idx=3, at=0.32}, {idx=4, at=0.39}, {idx=5, at=0.7}, {idx=6, at=0.88}, {idx=7, at=1.0}]`

But without the projection I get values matching yours:
`[{idx=1, at=0.0}, {idx=2, at=0.17}, {idx=3, at=0.35}, {idx=4, at=0.43}, {idx=5, at=0.72}, {idx=6, at=0.89}, {idx=7, at=1.0}]`

I think that's a fair point about using the point indices especially re: precision, I'm not sure if the geo points will 100% line up with the connectors but I'll look into it.

**Reply by Azbesciak** [reply]
Ok, regarding distance - there is my fault.
I was using `Coordinate.distance` from JTS library (java). And they are calculating Euclidean distance. Sorry for that.

**Reply by Azbesciak** [reply]
Thank you for increasing the precision since the September release :)
